### PR TITLE
feat: google analytics

### PR DESCRIPTION
**Pull request by simboonlong**
## Summary

- add GA to magic demo
- only gonna be logging in event action and category, label and value not necessary

## Changes

- track page view
- track start
- track issue
- track file downloaded
- track file drop
- track user drop off

## Issues

- https://support.google.com/analytics/answer/10315383?hl=en&ref_topic=9303319
- before oct 2020, was Universal Analytics. now is GA4
- ga.js deprecated, to use gtag instead

### CODE REVIEW

**Comment by cavacado** [comment]
lets kiv this first, need to w8 for the revised flow due to the concurrency/cancellation troubles

File: package.json [hunk 1]
```diff
@@ -89,6 +89,7 @@
     "@testing-library/react-hooks": "^7.0.0",
     "@types/debug": "^4.1.5",
     "@types/file-saver": "^2.0.3",
+    "@types/gtag.js": "0.0.8",
```
Comment (cavacado) [inline package.json]: why pin the dependency though?

File: public/static/index.html [hunk 2]
```diff
@@ -24,6 +24,20 @@
...
+      window.dataLayer = window.dataLayer || [];
+      function gtag(){dataLayer.push(arguments);}
+      gtag('js', new Date());
+
+      gtag('config', 'G-13GYPPVD4Y', {
+        'allow_google_signals': false,
+        'allow_ad_personalization_signals': false,
+        'debug_mode': '<%= htmlWebpackPlugin.options.isDev %>'
+      });
+   </script>
```
Comment (cavacado) [inline public/static/index.html:37]: can we inject the environmental variables for `GA_MEASUREMENT_ID`?

also can define the gtag config there as well (define the config in analytics.js, import and inject in htmlwebpackplugin)

File: src/common/analytics/Analytics.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+export const gaEvent = (event: GaEventProps): void => {
+  validateGaEvent(event);
+  const { action, category, label, value } = event;
+  gtag("event", action, {
+    event_category: category,
+    event_label: label,
+    value: value,
+  });
+};
```
Comment (cavacado) [inline src/common/analytics/Analytics.ts:53]: i was thinking wehther is it too harsh to throw an error if the payload is not correct for both page view and event.

if we let it fail silently, our logs will not have the analytics but the site still can function
but if we let it explode then the whole app will have a hard dependency on google analytics

File: package.json (hunk 1 above)
Comment (simboonlong) [reply package.json]: strange, re-tried fresh install it auto-pin still. nvm i manually put caret

File: public/static/index.html (hunk 2 above)
Comment (simboonlong) [reply public/static/index.html:37]: ok

File: src/common/analytics/Analytics.ts (hunk 3 above)
Comment (simboonlong) [reply src/common/analytics/Analytics.ts:53]: ok

File: webpack.config.js [hunk 4]
```diff
@@ -9,6 +9,13 @@ const Mode = require("frontmatter-markdown-loader/mode");
 const IS_DEV = process.env.NODE_ENV === "development";
 const IS_PROD = !IS_DEV;
 
+const GA_MEASUREMENT_ID = "G-13GYPPVD4Y";
+const GA_CONFIG_OPTION = {
+  allow_google_signals: false,
+  allow_ad_personalization_signals: false,
+  debug_mode: IS_DEV,
+};
```
Comment (cavacado) [inline webpack.config.js]: will u be importing from config later? (after the other pr)

File: src/common/analytics/Analytics.ts (hunk 3 above)
Comment (cavacado) [reply src/common/analytics/Analytics.ts:53]: 👍

File: src/common/analytics/Analytics.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+interface GaPageViewProps {
+  action: GaActionDefault;
+}
+
+export const validatePageViewEvent = (event: GaPageViewProps): void => {
+  const { action } = event;
+  if (!action) console.error("Action is required");
+  if (action && typeof action !== "string") console.error("Action must be a string");
```
Comment (cavacado) [inline src/common/analytics/Analytics.ts:26]: actually thinking about this validator, is it required?
i mean, so far every call to gaPageView is determined by us, there isnt anyway to call it from the user's point of view right?

File: webpack.config.js (hunk 4 above)
Comment (simboonlong) [reply webpack.config.js]: yeah

File: src/common/analytics/Analytics.ts (hunk 5 above)
Comment (cavacado) [reply src/common/analytics/Analytics.ts:26]: as discussed, we will ignore this for now since opencerts has this kind of behavior

**Review by cavacado — APPROVED** [review]

**Pull request merged**